src/librbd/mirror/snapshot/CreatePrimaryRequest.cc

```diff
@@ -205,6 +205,14 @@ void CreatePrimaryRequest<I>::unlink_peer() {
         unlink_snap_id = 0;
         continue;
       }
+      if (info->mirror_peer_uuids.size() == 0) {
+        peer_uuid = peer;
+        snap_id = snap_it.first;
+        break;
+      }
```

Comment:
I need to confirm but I think when you enable the mirroring on the image with the snapshot mode and there is no peer you can end up in a situation like this. There is also a similar check in UnlinkPeerRequest for this https://github.com/ceph/ceph/blob/master/src/librbd/mirror/snapshot/UnlinkPeerRequest.cc#L193 introduced in https://github.com/ceph/ceph/pull/3996. Maybe there would be some weird case if the librbd client (ie: the mgr) is abruptly stopped.

Reply:
A race possible, when `UnlinkPeerRequest` request is called simultaneously for the same `snap_id` but different `peer` (by two mirrors). They both may check that it is not the last peer and call "unlink_peer" cls (instead of the one of them, the last, calls snapshot remove). This race is handled in `UnlinkPeerRequest` by calling refresh after `unlink_peer`, to check if this situation happened (the peer list is empty) and remove the snapshot. But if the daemons crash at that time (or some error happen) we may end up with an "orphaned" snapshot.

So the additional check here looks useful.

Reply:
Thanks for the explanation!